From 789a0a950ac5cc76f320cb665d1106ce22ed7c69 Mon Sep 17 00:00:00 2001
From: Li Lingfeng <lilingfeng3@huawei.com>
Date: Tue, 10 May 2022 23:05:08 +0800
Subject: [PATCH] block: add info when opening an exclusive opened block device
 for write
Patch-mainline: Not yet, from openEuler
References: bsn#22
openEuler-commit: 789a0a950ac5cc76f320cb665d1106ce22ed7c69
Modified-by-SEL: No


hulk inclusion
category: feature
bugzilla: https://gitee.com/openeuler/kernel/issues/I53Q6M
CVE: NA

---------------------------

Just like open a write opend block device exclusively, open an exclusive
opened block device for write may also lead to potential data corruption.
This patch add an info message when opening an exclusive opened block
device for write to hint the potential data corruption.

Note that there are some legal cases such as file system or device
mapper online resize, so this message is just a hint and isn't always
mean that a risky written happens.

Signed-off-by: Li Lingfeng <lilingfeng3@huawei.com>
Reviewed-by: Zhang Yi <yi.zhang@huawei.com>
Reviewed-by: zhihao Cheng <chengzhihao1@huawei.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 fs/block_dev.c | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/fs/block_dev.c b/fs/block_dev.c
index 915d3b5bdee7..c8a3c93cc256 100644
--- a/fs/block_dev.c
+++ b/fs/block_dev.c
@@ -1651,6 +1651,18 @@ static int __blkdev_get(struct block_device *bdev, fmode_t mode, void *holder,
 				"block device exclusively");
 		bd_finish_claiming(bdev, claiming, holder);
 		spin_unlock(&bdev_lock);
+	} else if (!for_part && (mode & FMODE_WRITE)) {
+		spin_lock(&bdev_lock);
+		/*
+		 * Open an exclusive opened device for write may
+		 * probability corrupt the device, such as a
+		 * mounted file system, give a hint here.
+		 */
+		if (bdev->bd_holders ||
+		   (whole && (whole->bd_holder != NULL) && (whole->bd_holder != bd_may_claim)))
+			blkdev_dump_conflict_opener(bdev, "VFS: Open an exclusive opened "
+				"block device for write");
+		spin_unlock(&bdev_lock);
 	}
 
 	/*
-- 
2.33.0

